### PR TITLE
feat(platform/bitbucket): token authentication

### LINKED ISSUE #14221: Bitbucket cloud: allow using OAuth to token
### What would you like Renovate to be able to do?

I managed to set up a self-hosted instance of Renovate to check all our Bitbucket repositories, it's launched as a Bitbucket pipelines job.

However, at the moment it has to authenticate using an app password as a Bitbucket user. Meanwhile, Bitbucket allows [authenticating as an integration](https://developer.atlassian.com/cloud/bitbucket/rest/intro/) as well, when calling its API. After authenticating this way, the API requires the authorization header to supply a Bearer token, like this:

curl -X GET --url 'https://api.bitbucket.org/2.0/repositories/ORGNAME' --header 'Authorization: Bearer BEARERTOKEN' --header 'Accept: application/json'

I looked through Renovate's documentation (as well as the source code, but just a little), and it seems that this scenario is currently not supported. Instead, I have to use user and password to authenticate against Bitbucket API. Thus I would like to ask you to please add support Bearer auth as well.

### If you have any ideas on how this should be implemented, please tell us here.

Not sure how many changes would be needed, but I suspect not a lot. I tried querying the API with a bearer token issued this way, and I could get both the repository list and the user (=organization) profile the same way you are getting it.

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by nkuik**
## Changes

This PR allows authentication into Bitbucket via an auth token (as described in [this issue](https://github.com/renovatebot/renovate/issues/14221). It allows the usage of a token when initializing the Bitbucket platform, in addition to initializing a repo, provided that the token is available.

## Context

Closes #14221

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

I haven't actually done either of these 😱, as it wasn't completely apparent to me where exactly the auth token being used in the code changes of this PR was actually coming from. [BitBucket docs](https://developer.atlassian.com/cloud/bitbucket/rest/intro/#authentication) specify 4 methods for retrieving an access token, using one of two endpoints and passing a `client_id` and `secret` in the request. This token can then be used to make [subsequent API requests](https://developer.atlassian.com/cloud/bitbucket/rest/intro/#making-requests).

The issue creator (@rimas-kudelis) didn't seem to explain the details regarding how the auth token would be retrieved, and without knowing their setup, I cannot really determine how it happens.

I am happy to update the PR code and/or the docs if/when I receive additional context.

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [X] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/platform/bitbucket/index.ts [hunk 1]
```diff
@@ -68,13 +70,18 @@ export async function initPlatform({
   }
   setBaseUrl(defaults.endpoint);
   renovateUserUuid = null;
+  const options: InternalHttpOptions = {
```
Comment (nkuik) [inline lib/modules/platform/bitbucket/index.ts:73]: The `options` object can also be instantiated without the `InternalHttpOptions` type, so I don't know which way is preferred.

File: lib/modules/platform/bitbucket/index.spec.ts [hunk 2]
```diff
@@ -95,6 +95,15 @@ describe('modules/platform/bitbucket/index', () => {
       ).toMatchSnapshot();
     });
 
+    it('should init with only token', async () => {
+      httpMock.scope(baseUrl).get('/2.0/user').reply(200);
+      expect(
+        await bitbucket.initPlatform({
+          token: 'token',
+        })
+      ).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.spec.ts:104]: Use `toEqual` matcher instead. Don't follow other tests and read our best practices guid 🙃 

https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#unit-testing

File: lib/modules/platform/bitbucket/index.ts [hunk 3]
```diff
@@ -68,13 +70,18 @@ export async function initPlatform({
...
+    useCache: false,
+  };
+  if (token) {
+    options.headers = { Authorization: `Bearer ${token}` };
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:77]:
```suggestion
    options.token = token;
```

our http api should handle the header for you

File: lib/modules/platform/bitbucket/index.ts [hunk 4]
```diff
@@ -192,11 +199,14 @@ export async function initRepo({
   // TODO #7154
   const hostnameWithoutApiPrefix = regEx(/api[.|-](.+)/).exec(hostname!)?.[1];
 
+  const auth: string = opts.token
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:202]:
```suggestion
  const auth = opts.token
```

typescript shoud be smart enough 🙃

File: lib/modules/platform/bitbucket/index.ts [hunk 5]
```diff
@@ -192,11 +199,14 @@ export async function initRepo({
...
+    ? `x-token-auth:${opts.token}`
+    : // TODO: types (#7154)
+      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+      `${opts.username}:${opts.password}`;
   const url = git.getUrl({
     protocol: 'https',
-    // TODO: types (#7154)
-    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
-    auth: `${opts.username}:${opts.password}`,
+    auth: auth,
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:209]:
```suggestion
    auth,
```

File: lib/modules/platform/bitbucket/index.ts [hunk 6]
```diff
@@ -192,11 +199,14 @@ export async function initRepo({
   // TODO #7154
   const hostnameWithoutApiPrefix = regEx(/api[.|-](.+)/).exec(hostname!)?.[1];
 
+  const auth: string = opts.token
+    ? `x-token-auth:${opts.token}`
+    : // TODO: types (#7154)
+      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+      `${opts.username}:${opts.password}`;
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:206]:
```suggestion
      `${opts.username!}:${opts.password!}`;
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/platform/bitbucket/index.ts (hunk 4 above)
Comment (nkuik) [reply lib/modules/platform/bitbucket/index.ts:202]: The `token` property doesn't exist on the [InternalHttpOptions interface](https://github.com/renovatebot/renovate/blob/main/lib/util/http/types.ts#L70). Let me know if I should take a different strategy

File: lib/modules/platform/bitbucket/index.ts (hunk 3 above)
Comment (nkuik) [reply lib/modules/platform/bitbucket/index.ts:78]: The `token` property doesn't seem to exist on the [InternalHttpOptions interface](https://github.com/renovatebot/renovate/blob/main/lib/util/http/types.ts#L70). Let me know if I should take a different strategy

File: lib/modules/platform/bitbucket/index.spec.ts [hunk 7]
```diff
@@ -3,9 +3,9 @@ import type { logger as _logger } from '../../../logger';
 import { BranchStatus, PrState } from '../../../types';
 import type * as _git from '../../../util/git';
 import { setBaseUrl } from '../../../util/http/bitbucket';
-import type { Platform, RepoParams } from '../types';
+import type { Platform, PlatformResult, RepoParams } from '../types';
 
-const baseUrl = 'https://api.bitbucket.org';
+const baseUrl = 'https://api.bitbucket.org/';
```
Comment (nkuik) [inline lib/modules/platform/bitbucket/index.spec.ts:8]: Not sure exactly why this URL lacks the trailing `/` when the other places it's defined (like [here](https://github.com/renovatebot/renovate/blob/main/lib/modules/platform/bitbucket/index.ts#L43) and [here](https://github.com/renovatebot/renovate/blob/main/lib/util/http/bitbucket.ts#L5) it includes the trailing `/`

**Comment by rimas-kudelis** [comment]
@nkuik, it's been almost six months since I made the feature request, so I don't really remember which exact way I used to get an OAuth token. Nor should it matter, I think.

In the settings of a BitBucket workspace, you can find the "OAuth consumers" item, where new consumers can be registered. I think this is how I created a consumer back then, which, once created, got a key and a secret. What I wanted was to be able to give this key and secret to Renovate instead of a user and password, and for it to work. I wouldn't really care about which particular method of getting a token Renovate would use as long as the integration worked.

File: lib/modules/platform/bitbucket/index.spec.ts (hunk 7 above)
Comment (viceice) [reply lib/modules/platform/bitbucket/index.spec.ts:8]: revert, we shoud generally don't use trailing slashes. most of the time we trim them anyways

File: lib/modules/platform/bitbucket/index.spec.ts [hunk 8]
```diff
@@ -85,14 +85,29 @@ describe('modules/platform/bitbucket/index', () => {
...
+    });
+
+    it('should init with only token', async () => {
+      const expectedResult: PlatformResult = {
+        endpoint: baseUrl,
+      };
+      httpMock.scope(baseUrl).get('/2.0/user').reply(200);
+      expect(
+        await bitbucket.initPlatform({
+          token: 'token',
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.spec.ts:108]:
```suggestion
          token: 'abc',
```

needs to be shorter to pass codeql

File: lib/modules/platform/bitbucket/__snapshots__/index.spec.ts.snap [hunk 9]
```diff
@@ -150,6 +150,18 @@ Object {
...
+  "endpoint": "https://api.bitbucket.org/",
+}
+`;
+
+exports[`modules/platform/bitbucket/index initPlatform() should init with username/password 1`] = `
+Object {
+  "endpoint": "https://api.bitbucket.org/",
+}
+`;
+
```
Comment (viceice) [inline lib/modules/platform/bitbucket/__snapshots__/index.spec.ts.snap:164]: remove

File: lib/modules/platform/bitbucket/index.ts [hunk 10]
```diff
@@ -8,6 +8,7 @@ import { BranchStatus, PrState, VulnerabilityAlert } from '../../../types';
 import * as git from '../../../util/git';
 import * as hostRules from '../../../util/host-rules';
 import { BitbucketHttp, setBaseUrl } from '../../../util/http/bitbucket';
+import type { InternalHttpOptions } from '../../../util/http/types';
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:11]:
```suggestion
import type { HttpOptions } from '../../../util/http/types';
```

File: lib/modules/platform/bitbucket/index.ts (hunk 1 above)
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:73]:
```suggestion
  const options: HttpOptions = {
```

File: lib/modules/platform/bitbucket/index.ts (hunk 3 above)
Comment (viceice) [reply lib/modules/platform/bitbucket/index.ts:77]: I'll add it to `HttpOptions` interface on seperate PR

- #16967

**Review by viceice — CHANGES_REQUESTED** [review]

**Review by viceice — DISMISSED** [review]

File: lib/modules/platform/bitbucket/index.spec.ts [hunk 11]
```diff
@@ -85,14 +85,31 @@ describe('modules/platform/bitbucket/index', () => {
       );
     });
 
-    it('should init', async () => {
+    it('should init with username/password', async () => {
+      const expectedResult: PlatformResult = {
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.spec.ts:89]: please inline this small object

File: lib/modules/platform/bitbucket/index.spec.ts [hunk 12]
```diff
@@ -85,14 +85,31 @@ describe('modules/platform/bitbucket/index', () => {
...
+          endpoint: baseUrl,
           username: 'abc',
           password: '123',
         })
-      ).toMatchSnapshot();
+      ).toEqual(expectedResult);
+    });
+
+    it('should init with only token', async () => {
+      const expectedResult: PlatformResult = {
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.spec.ts:103]: same here

File: lib/modules/platform/bitbucket/index.ts [hunk 13]
```diff
@@ -55,10 +56,11 @@ export async function initPlatform({
   endpoint,
   username,
   password,
+  token,
 }: PlatformParams): Promise<PlatformResult> {
-  if (!(username && password)) {
+  if (!(username && password) && !token) {
     throw new Error(
-      'Init: You must configure a Bitbucket username and password'
+      'Init: You must either configure a Bitbucket token or username and password'
```
Comment (rarkins) [inline lib/modules/platform/bitbucket/index.ts:63]:
```suggestion
      'Init: You must configure either a Bitbucket token or username and password'
```

**Review by rarkins — APPROVED** [review]

**Pull request merged**